logstash-core/lib/logstash/instrument/metric.rb

```diff
@@ -0,0 +1,109 @@
+# encoding: utf-8
+require "logstash/instrument/collector"
+require "concurrent"
+
+module LogStash module Instrument
+  class MetricException < Exception; end
+  class MetricNoKeyProvided < MetricException; end
+  class MetricNoBlockProvided < MetricException; end
+  class MetricNoNamespaceProvided < MetricException; end
+
+  # This class provide the interface between the code, the collector and the format
+  # of the recorded metric.
+  class Metric
```

Comment:
Conceptually I find the use of this Metric class a bit strange.
It's possible to create a generic Metric with `metric = Metric.create` but then increment/decrement/gauge multiple different metrics with `metric.increment(:pipeline, :in, 1); metric.decrement(:pipeline, :out, 2)`.

This generic metric is essentially a stateless proxy to the collector with extra validation. Using this generic metric is a bit confusing when reading the code. The agent initalizer uses `configure_metric` to create a single metric for itself, when you'd expect it to have many. It's actually not a metric at all, but just the gateway towards the collector.

Thoughts?